extensions/amp-a4a/0.1/amp-a4a.js

```diff
@@ -744,6 +792,47 @@ export class AmpA4A extends AMP.BaseElement {
   }
 
   /**
+   * Refreshes ad slot by fetching a new creative and rendering it. This leaves
+   * the current creative displayed until the next one is ready.
+   *
+   * @param {function()} refreshEndCallback When called, this function will
+   *   restart the refresh cycle.
+   */
+  refresh(refreshEndCallback) {
+    this.isRefreshing = true;
+    this.tearDownSlot();
+    this.initiateAdRequest();
+    const promiseId = this.promiseId_;
+    this.adPromise_.then(() => {
+      if (!this.isRefreshing || promiseId != this.promiseId_) {
+        // If this refresh cycle was canceled, such as in a no-content
+        // response case, keep showing the old creative.
+        refreshEndCallback();
+        return;
+      }
+      this.togglePlaceholder(true);
```

Comment:
Unfortunately, I need to call refreshEndCallback(), which is passed to refresh() from the RefreshManager, after rendering is complete to restart the refresh cycle. I can move all of the DOM mutating stuff to layoutCallback, but I'll still need refreshReadyPromise_ so that I can signal within refresh() when to call refreshEndCallback(). Also, timerFor(this.win).delay() returns a cancellation id, not a promise; not sure how much that matters.